source/core_http_client.c

```diff
@@ -1706,55 +1740,77 @@ HTTPStatus_t HTTPClient_AddRangeHeader( HTTPRequestHeaders_t * pRequestHeaders,
 /*-----------------------------------------------------------*/
 
 static HTTPStatus_t sendHttpData( const TransportInterface_t * pTransport,
+                                  HTTPClient_GetCurrentTimeFunc_t getTimestampMs,
                                   const uint8_t * pData,
                                   size_t dataLen )
 {
     HTTPStatus_t returnStatus = HTTPSuccess;
     const uint8_t * pIndex = pData;
-    int32_t transportStatus = 0;
+    int32_t bytesSent = 0;
     size_t bytesRemaining = dataLen;
+    uint32_t lastSendTimeMs = 0U, timeSinceLastSendMs = 0U;
+    uint32_t retryTimeoutMs = HTTP_SEND_RETRY_TIMEOUT_MS;
 
     assert( pTransport != NULL );
     assert( pTransport->send != NULL );
     assert( pData != NULL );
 
+    /* If the timestamp function was undefined by the application, then do not
+     * retry the transport send. */
+    if( getTimestampMs == getZeroTimestampMs )
+    {
+        retryTimeoutMs = 0U;
+    }
+
+    /* Start the last send time to allow retries on the first zero data sent. */
+    lastSendTimeMs = getTimestampMs();
+
     /* Loop until all data is sent. */
     while( ( bytesRemaining > 0UL ) && ( returnStatus != HTTPNetworkError ) )
     {
-        transportStatus = pTransport->send( pTransport->pNetworkContext,
-                                            pIndex,
-                                            bytesRemaining );
+        bytesSent = pTransport->send( pTransport->pNetworkContext,
+                                      pIndex,
+                                      bytesRemaining );
 
         /* A transport status of less than zero is an error. */
```

Comment:
The variable is now called `bytesSent`, so this comment is less consistent now.

Reply:
Sure